src/components/gui/trigger-editor/index.tsx

```diff
@@ -0,0 +1,193 @@
+import { useDatabaseDriver } from "@/context/driver-provider";
+import { DatabaseTriggerSchemaChange, TriggerOperation, TriggerWhen } from "@/drivers/base-driver";
+import { LucideAlertCircle } from "lucide-react";
+import { useCallback, useEffect, useMemo, useState } from "react";
+import TableCombobox from "../table-combobox/TableCombobox";
+import { Input } from "@/components/ui/input";
+import { Select, SelectContent, SelectItem, SelectTrigger, SelectValue } from "@/components/ui/select";
+import SqlEditor from "../sql-editor";
+import { TriggerController } from "./trigger-controller";
+import { TriggerSaveDialog } from "./trigger-save-dialog";
+
+export interface TriggerEditorProps {
+  name: string;
+  tableName?: string;
+  schemaName: string;
+}
+
+interface Props extends TriggerEditorProps {
+  onSave: (trigger: TriggerEditorProps) => void;
+}
+
+const initailTrigger: DatabaseTriggerSchemaChange = {
+  name: {
+    new: ''
+  },
+  operation: "INSERT",
+  when: "BEFORE",
+  tableName: "",
+  whenExpression: "",
+  statement: "",
+  schemaName: ''
+}
+
+function triggerHasChange(defaultTrigger: DatabaseTriggerSchemaChange, trigger: DatabaseTriggerSchemaChange) {
+  const objects = Object.keys(defaultTrigger) as (keyof DatabaseTriggerSchemaChange)[];
+
+  for (const key of objects) {
+    if (defaultTrigger[key as keyof DatabaseTriggerSchemaChange] !== trigger[key as keyof DatabaseTriggerSchemaChange]) {
+      return true;
+    }
+  }
+
+  return false;
+}
+
+export default function TriggerEditor(props: Props) {
+  const { name, tableName, schemaName } = props;
+  const { databaseDriver } = useDatabaseDriver();
+  const [defaultTrigger, setDefaultTrigger] = useState<DatabaseTriggerSchemaChange>({
+    ...initailTrigger
+  })
+  const [trigger, setTrigger] = useState<DatabaseTriggerSchemaChange>({
+    ...initailTrigger
+  });
```

Comment:
Try to use

https://immerjs.github.io/immer/
https://immerjs.github.io/immer/example-setstate

This will simpilfy the code a lot of more. I will convert some my code in the future to immer as well